src/zarr/storage/_cache.py

```diff
@@ -0,0 +1,735 @@
+        if value_size <= self._max_size:
+            self._accommodate_value(value_size)
+            cache_key = self._normalize_key(key)
+            self._values_cache[cache_key] = cache_value
+            self._current_size += value_size
+        # If value_size > max_size, we simply don't cache it (silent skip)
+
+    def invalidate(self) -> None:
+        """Completely clear the cache."""
+        with self._mutex:
+            self._values_cache.clear()
+            self._invalidate_keys()
+            self._current_size = 0
+
+    def invalidate_values(self) -> None:
+        """Clear the values cache."""
+        with self._mutex:
+            self._values_cache.clear()
+
+    def invalidate_keys(self) -> None:
+        """Clear the keys cache."""
+        with self._mutex:
+            self._invalidate_keys()
+
+    def _invalidate_keys(self) -> None:
+        self._keys_cache = None
+        self._contains_cache.clear()
+        self._listdir_cache.clear()
+
+    def _invalidate_value(self, key: Any) -> None:
+        cache_key = self._normalize_key(key)
+        if cache_key in self._values_cache:
+            value = self._values_cache.pop(cache_key)
+            self._current_size -= buffer_size(value)
+
+    def __getitem__(self, key: Any) -> Any:
+        cache_key = self._normalize_key(key)
+        try:
+            # first try to obtain the value from the cache
+            with self._mutex:
+                value = self._values_cache[cache_key]
+                # cache hit if no KeyError is raised
+                self.hits += 1
+                # treat the end as most recently used
+                self._values_cache.move_to_end(cache_key)
+
+        except KeyError:
+            # cache miss, retrieve value from the store
+            if hasattr(self._store, "__getitem__"):
+                value = self._store[key]
+            else:
+                # Fallback for async stores
+                raise KeyError(f"Key {key} not found in store") from None
+            with self._mutex:
+                self.misses += 1
+                # need to check if key is not in the cache, as it may have been cached
+                # while we were retrieving the value from the store
+                if cache_key not in self._values_cache:
+                    self._cache_value(cache_key, value)
+
+        return value
+
+    def __setitem__(self, key: str, value: Buffer) -> None:
+        if hasattr(self._store, "__setitem__"):
+            self._store[key] = value
+        else:
+            # For async stores, we can't handle this synchronously
+            raise TypeError("Cannot use __setitem__ with async store")
+
+        # Update cache and invalidate keys cache since we may have added a new key
+        with self._mutex:
+            self._invalidate_keys()
+            self._cache_value(self._normalize_key(key), value)
+
+    def __delitem__(self, key: Any) -> None:
```

Comment:
Why implement these special methods (including `__setitem__`? No other stores seem to implement them, and have other methods that do the same thing instead.